File: leetcode/medium/0098ValidateBST/valid.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        return self.recursive_check(root)
    def recursive_check(self, node, lower = float('-inf'), upper = float('inf')):
        if not node:
            return True
        
        val = node.val
        if val <= lower or val >= upper:
            return False

        if not self.recursive_check(node.right, val, upper):
            return False
        if not self.recursive_check(node.left, lower, val):
            return False
        return True

File: leetcode/medium/0098ValidateBST/valid.cpp
#include "bits/stdc++.h";
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};



// Accepted solution but because cpp is statically typed, there might be integer overflow
class RecursiveSolution
{
public:
    bool isValidBST(TreeNode *root)
    {

        return isValidBST(root, INT_MIN + 1 , INT_MAX - 1);
    }

private:
    bool isValidBST(TreeNode *node, int min, int max)
    {

        if (node == NULL)
            return true;

        if (node->val < min || node->val > max)
            return false;

        if (node->left != NULL && node->left->val >= node->val)
            return false;

        if (node->right != NULL && node->right->val <= node->val)
            return false;

        return isValidBST(node->left, min, node->val - 1) && isValidBST(node->right, node->val + 1, max);
    }
};
